joblib/parallel.py

```diff
@@ -268,6 +273,72 @@ def effective_n_jobs(n_jobs=-1):
     return backend.effective_n_jobs(n_jobs=n_jobs)
 
 
+def choose_backend_and_n_jobs(backend, n_jobs, prefer_processes=True,
+                              require_shared_memory=False):
+    """Determine the backend and n_jobs to use for a ``Parallel`` call.
+
+    Parameters
+    ----------
+    backend : str, ParallelBackendBase instance or None
+        The backend specified in the call to Parallel, if any
+    n_jobs : int
+        The number of jobs specified in the call to Parallel, if any
+    prefer_processes : bool, optional
+        If true, backends using multiple processes are prefered to single
+        process backends (e.g. threading). This is useful if the function
+        being applied doesn't release the global interpretter lock (GIL).
+    require_shared_memory : bool, optional
+        If true, a backend supporting shared memory (e.g. a single process) is
+        required. If no valid backend option exists, a ``ValueError`` will be
+        raised.
+
+    Returns
+    -------
+    backend : ParallelBackendBase
+        The backend to use
+    """
+    # Check for compatible preferences and requiremements
+    if require_shared_memory and prefer_processes:
+        raise ValueError("Backends with multiple processes cannot have "
+                         "shared memory")
+
+    # Determine the specified backend in the call to `Parallel`
+    if backend is None:
+        default_backend = BACKENDS[DEFAULT_BACKEND]()
+    elif isinstance(backend, ParallelBackendBase):
+        default_backend = backend
+    else:
+        try:
+            backend_factory = BACKENDS[backend]
+        except KeyError:
+            raise ValueError("Invalid backend: %s, expected one of %r"
+                             % (backend, sorted(BACKENDS.keys())))
+        default_backend = backend_factory()
+
+    # If there is an active backend, prefer that
+    if has_active_backend():
+        backend_and_n_jobs = [get_active_backend(), (default_backend, n_jobs)]
+    else:
+        backend_and_n_jobs = [(default_backend, n_jobs)]
+
+    # If shared memory is required, use the first backend that supports it
+    if require_shared_memory:
+        for backend, n_jobs in backend_and_n_jobs:
+            if getattr(backend, 'uses_processes', None) is False:
+                return backend, n_jobs
+        raise ValueError("Unable to satisfy shared memory requirement "
```

Comment:
Looks like this case isn't hit by the test suite. 

This is a silly example, but does hit the code:

```python
choose_backend_and_n_jobs('multiprocessing', 5,
                          require_shared_memory=True, prefer_processes=False)
```